pkg/certificatesigningrequest/controller.go

```diff
@@ -0,0 +1,168 @@
+// Package certificatesigningrequest implements the controller for Node Certificate Signing Request.
+package certificatesigningrequest
+
+import (
+	"context"
+	"crypto/x509"
+	"fmt"
+	"strings"
+	"time"
+
+	certificatesv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8swatch "k8s.io/apimachinery/pkg/watch"
+	clientkubernetes "k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+)
+
+// ProviderChecks is a function that checks if the CertificateSigningRequest is valid in the provider.
+type ProviderChecks func(context.Context, clientkubernetes.Interface, *x509.CertificateRequest) (bool, error)
+
+// Reconciler is the controller for CertificateSigningRequest.
+type Reconciler struct {
+	kclient        clientkubernetes.Interface
+	providerChecks ProviderChecks
+}
+
+// NewCsrController returns a new CertificateSigningRequest controller.
+func NewCsrController(kclient clientkubernetes.Interface, fn ProviderChecks) *Reconciler {
+	return &Reconciler{
+		kclient:        kclient,
+		providerChecks: fn,
+	}
+}
+
+// Run the CertificateSigningRequest controller.
+//
+//nolint:gocyclo
+func (r *Reconciler) Run(ctx context.Context) {
+	watchTimeoutSeconds := int64(time.Minute * 5)
+
+	for {
+		watcher, err := r.kclient.
+			CertificatesV1().
+			CertificateSigningRequests().
+			Watch(ctx, metav1.ListOptions{
+				Watch:          true,
+				TimeoutSeconds: &watchTimeoutSeconds, // Default timeout: 20 minutes.
+			})
+		if err != nil {
+			klog.Errorf("CertificateSigningRequestReconciler: failed to list CSR resources: %v", err)
+			time.Sleep(10 * time.Second) // Pause for a while before retrying, otherwise we'll spam error logs.
```

Comment:
doesn;t the controller runtime has a way to return a backoff rather than doing a sleep?

Reply:
Sorry, i did not get it. Can you provide more details?

During the watch process, we can get reset or connection timeout. I do not want to retry immediately, only after a few second.